### PR TITLE
Feature/te 309 allow assertions

### PR DESCRIPTION
**Pull request by gunther-bachmann**

### CODE REVIEW

File: tcl/org.testeditor.tcl.model/model/tcl.xcore [hunk 1]
```diff
@@ -51,8 +51,21 @@ class TestStep {
 	contains StepContent[0..*] contents
 }
 
+class TestStepAssertion extends TestStep {
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.model/model/tcl.xcore]: I would prefer `AssertionTestStep`, this is less ambiguous. It is a test step, that performs an assertion.

File: tcl/org.testeditor.tcl.model/model/tcl.xcore [hunk 2]
```diff
@@ -51,8 +51,21 @@ class TestStep {
 	contains StepContent[0..*] contents
 }
 
+class TestStepAssertion extends TestStep {	
+	String variableName
+	contains Comparator comparator
+	Boolean negated
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.model/model/tcl.xcore]: can be `boolean`, default is `false`

File: tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/jvmmodel/TclJvmModelInferrer.xtend [hunk 3]
```diff
@@ -38,7 +39,8 @@ class TclJvmModelInferrer extends AbstractModelInferrer {
 	@Inject extension JvmTypesBuilder
 	@Inject extension TclModelUtil
 	@Inject IQualifiedNameProvider nameProvider
-	
+	@Inject extension TclAssertCallBuilder
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/jvmmodel/TclJvmModelInferrer.xtend]: I like the extraction in a separate class, but I would remove the `extension` and specify the name to `assertBuilder`

File: tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext [hunk 4]
```diff
@@ -51,6 +49,22 @@ TestStep:
 	'-' contents+=StepContent* '.'?
 ;
 
+TestStepAssertion:
+	'-' 'assert' (negated?='!')? variableName=ID (comparator=Comparator contents+=StepContent)? '.'?
+;
+
+Comparator:
+	{ComparatorEquals} '=' |
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext]: `=` shouldn't be the equals operator

File: tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/jvmmodel/TclJvmModelInferrer.xtend [hunk 5]
```diff
@@ -100,9 +102,11 @@ class TclJvmModelInferrer extends AbstractModelInferrer {
 		return fixtureType.simpleName.toFirstLower
 	}
 
-	private def void generateFeatureCall(TestStep step, ITreeAppendable output) {
-		output.newLine
-		val context = step.context
+	private def dispatch void toUnitTestCodeLine(TestStepAssertion step, ITreeAppendable output) {
+		output.append(buildAssertCall(step.negated, step.variableName, step.comparator, step.contents.restoreString)).newLine
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/jvmmodel/TclJvmModelInferrer.xtend]: Can we extract the call to `buildAssertCall`to a separate variable?

File: tcl/org.testeditor.tcl.model/model/tcl.xcore [hunk 6]
```diff
@@ -51,8 +51,21 @@ class TestStep {
...
+}
+
 class TestStepWithAssignment extends TestStep {
 	String variableName
 }
 
-class StepContentElement extends StepContent {}
\ No newline at end of file
+class StepContentElement extends StepContent {}
+class Comparator extends StepContent {
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.model/model/tcl.xcore]: Why is this a subclass of `StepContent`?

File: tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext [hunk 7]
```diff
@@ -51,6 +49,22 @@ TestStep:
 	'-' contents+=StepContent* '.'?
 ;
 
+TestStepAssertion:
+	'-' 'assert' (negated?='!')? variableName=ID (comparator=Comparator contents+=StepContent)? '.'?
+;
+
+Comparator:
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext]: This rule should be split into separate rules.

File: tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext [hunk 8]
```diff
@@ -51,6 +49,22 @@ TestStep:
 	'-' contents+=StepContent* '.'?
 ;
 
+TestStepAssertion:
+	'-' 'assert' (negated?='!')? variableName=ID (comparator=Comparator contents+=StepContent)? '.'?
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext]: I would see the `!` as part of the not-null operator only - or some future more powerful expressions like `!(...)`. Right now, it doesn't make sense to write `!x == "something"` and might be confusing. Does it mean `!(x == "something")` or does it mean `(!x) == "something"`?

Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext]: `contents+=StepContent` is too generic as it allows: `StepContentText | StepContentElement | StepContentVariable`

File: tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/jvmmodel/TclAssertCallBuilder.xtend [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+			ComparatorGreaterThen: if (comparator.negated ?: false) {
+				return '<='
+			} else {
+				return '>'
+			}
+			default: throw new RuntimeException('''comparator class «comparator.class» cannot produce order comparator''')
+		}
+	}
+
+	public def String buildAssertCall(Boolean assertNegate, String variableName, Comparator comparator,
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/jvmmodel/TclAssertCallBuilder.xtend]: Why not pass the whole test step here?

File: tsl/org.testeditor.tsl.dsl.web/webapp/xtext-resources/generated/mode-tsl.js [hunk 10]
```diff
@@ -1,24 +1,24 @@
-define(["ace/lib/oop", "ace/mode/text", "ace/mode/text_highlight_rules"], function(oop, mText, mTextHighlightRules) {
```
Comment (gunther-bachmann) [inline tsl/org.testeditor.tsl.dsl.web/webapp/xtext-resources/generated/mode-tsl.js]: This has sneaked in (no change actually done)

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/parser/TclModelParserTest.xtend [hunk 11]
```diff
@@ -93,26 +97,31 @@ class TclModelParserTest extends AbstractParserTest {
...
-				componentNode.text.assertEquals('GreetingsApplication')
-				steps.assertSize(2)
-				steps.get(0) => [
-					contents.restoreString.assertEquals('starte Anwendung "org.testeditor.swing.exammple.Greetings"')	
-				]
-				steps.get(1) => [
-					contents.restoreString.assertEquals('gebe in <Eingabefeld> den Wert "Hello World" ein')
-				]
+		test.steps.assertSingleElement =>
+			[
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/parser/TclModelParserTest.xtend]: Formatting seems broken here, was it the auto-formatter?

File: tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext [hunk 12]
```diff
@@ -12,81 +12,111 @@
...
+
+AssertionTestStep:
+	'-' 'assert' expression=AssertionExpression '.'?;
+
+/** Assertion Exression is either a null check, 
+ *  or a full comparison that requires a comparator and a right expression (in order to have no ambiguity with the null check), 
+ *  or an inner expression in parenthesis (no more null checks allowed in here!) 
+ */
+AssertionExpression:
+	AENullCheck | AEFullComparison | '(' AssertionInnerExpression ')';
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext]: We should refactor the class names of the subclasses of `AssertionExpression`. But not necessarily now.

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/parser/TclModelParserTest.xtend (hunk 11 above)
Comment (gunther-bachmann) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/parser/TclModelParserTest.xtend]: yep, autoformatter goes crazy within this file

File: tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext (hunk 12 above)
Comment (gunther-bachmann) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/Tcl.xtext]: understood, naming is always a tough issue

**Pull request merged**